src/executor_utils.c

```diff
@@ -0,0 +1,155 @@
+/*                                                                            */
+/* ************************************************************************** */
+
+/**
+ * @file executor_utils.c
+ * @brief Helper functions for executor.
+ */
+
+#include "mod_executor.h"
+
+/**
+ * @brief Check given args executable.
+ * 
+ * If args executable contains an absolute path (starts with /),
+ * a relative path (starts with ./ or ../) or no paths are given,
+ * its validity is checked in the current directory. If not successfully,
+ * ERR_UNKNOWN_CMD is returned.
+ * 
+ * Else an attempt to find and assign the right path is started. If not
+ * successfully, ERR_UNKNOWN_CMD is returned.  
+ * @param args 			String array containing executable in first position.
+ * @param cmd_paths 	String array of paths.
+ * @return t_err 		ERR_UNKNOWN_CMD, ERR_MALLOC, SUCCESS
+ */
+t_err	ft_check_cmd_access(char **args, char **cmd_paths)
+{
+	t_err	err;
+
+	err = ERR_UNKNOWN_CMD;
+	if ((*args)[0] == '/' || !ft_strncmp(*args, "./", 2)
+		|| !ft_strncmp(*args, "../", 3) || !cmd_paths)
+	{
+		if (access(*args, F_OK | X_OK) == 0)
+			return (SUCCESS);
+	}
+	else
+		err = ft_prefix_path(args, cmd_paths);
+	return (err);
+}
+
+/**
+ * @brief Prefix the correct path before the executable.
+ * 
+ * Rotate through all paths and checks whether the combination
+ * with the executable is correct. If unsucessful, ERR_UNKNOWN_CMD is
+ * returned.
+ * @param args 			String array containing executable in first position.
+ * @param cmd_paths 	String array of paths.
+ * @return t_err 		ERR_UNKNOWN_CMD, ERR_MALLOC, SUCCESS
+ */
+t_err	ft_prefix_path(char **args, char **cmd_paths)
+{
+	char	*tmp;
+	char	*rtrn;
+
+	while (*cmd_paths)
+	{
+		tmp = ft_strjoin(*cmd_paths, "/");
+		rtrn = ft_strjoin(tmp, *args);
+		free(tmp);
+		if (!rtrn)
+			return (ERR_MALLOC);
+		if (access(rtrn, F_OK | X_OK) == 0)
+		{
+			free(*args);
+			*args = rtrn;
+			return (SUCCESS);
+		}
+		free(rtrn);
+		cmd_paths++;
+	}
+	return (ERR_UNKNOWN_CMD);
+}
+
+/**
+ * @brief Get the path object from envp str array.
+ * 
+ * Loop through the envp string array.
+ * If envp str is PATH, then 0 || 0 -> exit while loop.
+ * If envp str is PATH2, then 0 || 1 -> stay in while loop.
+ * If no PATH is found (*envp = NULL), return ERR_NOPATH.
+ * If PATH is empty, a '\0' is found and saved into *paths. This will
+ * be treated as ERR_NOPATH as well.
+ * @param envp 		String array containing all environment strings.
+ * @param paths		Pointer to string array containing all path strings.
+ * @return t_err	ERR_NOPATH, ERR_MALLOC, SUCCESS
+ */
+t_err	ft_get_path(char **envp, char ***paths)
+{
+	char	*path_str;
+
+	while (*envp && (ft_strncmp(*envp, "PATH", 4) || (*(*envp + 4) != '=')))
+		envp++;
+	if (!*envp)
+		return (ERR_NOPATH);
+	path_str = *envp + 5;
+	*paths = ft_split(path_str, ':');
+	if (!*paths)
+		return (ERR_MALLOC);
+	if (***paths == '\0')
```

Comment:
In theory yes, ft_split would have an empty string with '\0'. But that should never occur as PATH is constant and is never with an empty value between colons. And I do not think that the shell is supposed to be able to handle the case when someone willfully unsets and resets PATH to this form.

Reply:
PATH is not constant. You can add and remove stuff as you like, that's the purpose of PATH. 

Examples: 

- If you want to exe your little progs in your home directory you can add the home directory to PATH. Then just type the program name.
- If you want to little what users can just exe, you can remove some directories from PATH.
- If you want to trip someone during eval you can unset PATH and see what happens.

Since PATH is can be altered as you like it can also be 'wrong'. Question is what do we have to catch. Bash doesn't care if there are empty colons at the start.

Reply:
Now that is an interesting behaviour. In that case an empty string in paths will lead to paths being freed and its pointer set to NULL. When checking for the cmd accessibility that case will be caught by

if ((*args)[0] == '/' || !ft_strncmp(*args, "./", 2)  || !ft_strncmp(*args, "../", 3) || !cmd_paths)

ergo the last condition. That covers the behaviour of each cmd being directly executable in its own directory. If it is not in its directory, an error message type of "cmd not found" will be send. 

So basically the status quo is good.

Reply:
No that's still a problem: we can _only_ execute in the same directory. If you add an empty string to PATH it gives you the possibility to _also_ execute in current directory.
If exe is not there, check in other PATHs

We currently behave differently if empty string is 
at start (clear PATH) or 
in the middle or at the end (leave PATH including empty string)

I think the solution is simple: just not check if there is an empty string. If ft_split is success it returns an array (which is checked). If we have an array we don't care what the strings are. Could be twenty empty strings and /usr/lib - we just check all of them.

Reply:
Let me get this straight: 

If there is an empty string in PATH -> 
1) we can execute the cmd, if we're in the currently in the same directory (no need for ./)
2) if we are not in the same directory, it will iterate through PATH and try to find the right directory: if its directory is not in PATH, it will display an error. If its directory is in PATH, it will execute

Is that the behaviour you're describing? If yes, that would mean that the code goes into both cases of ft_check_cmd_access():

	if ((*args)[0] == '/' || !ft_strncmp(*args, "./", 2)
		|| !ft_strncmp(*args, "../", 3) || !cmd_paths)
	{
		if (access(*args, F_OK | X_OK) == 0)
			return (SUCCESS);
	}
	else
		err = ft_prefix_path(args, cmd_paths);

Do both, first the access check with the cmd in the current directory and if that's not successfull, try to prefix the right path.

Reply:
Not exactly. The behavior is just "iterate through PATH". Let's say PATH is /usr/bin/:/usr/sbin/: which should split into
1. "/usr/bin/"
2. "/usr/sbin/"
3. ""

After strjoin with cmd we check access of
1. /usr/bin/cmd
2. /usr/sbin/cmd
3. cmd

If we have no path we would also check the same as 3, but that's just because we can't append anything.

Reply:
Yes it does have the same effect but we don't even need to check anything extra.

Why do a directory check if everything works if we just delete the if condition?

What is the purpose of the if condition in lines 109-114?

Reply:
You would need to check the whole array since an empty string can be anywhere and more than once.

I.e. :::/usr/lib/:::

Why should this happen? Because eval.

Alternative: if empty string: don't add '/'

Reply:
Had a closer look at ft_split(): It doesn't create empty strings in case of ::, it just jumps over these instances.

In order to catch this we need to write an extra function that sets a flag in case it encounters double colons in the PATH. Only then we can handle this special behaviour.

Reply:
That sucks :( 

Other special cases:

- PATH=: (just colon also works)
- PATH=/usr/lib/: (adding just one colon at the end also works) 

How does ft_split handle PATH="" (just empty string)?

Reply:
Need to check these cases individually to be sure, but I think:
PATH=: creates an array with just NULL inside
PATH=/usr/lib/: creates an array with /usr/lib/ and NULL
PATH="" creates an array with just NULL inside

What is the effect of these cases on bash behaviour that we need to imitate?

Reply:
All three make it possible to execute exe in current directory. 

I think you have to see it as empty string equals exe in current directory (since cmd + "" = cmd):

PATH=: equals two PATHs (left and right side of colon). Both are empty string.
PATH=/usr/lib/: equals two PATHs. Right side of colon is empty string.
PATH="" equals one PATH. Empty string.

Reply:
Conferenced - looks good!